File: src/token_bucket.py
import time
from threading import Lock


class TokenBucket:
    def __init__(self, max_tokens: float, refill_rate: float):
        assert max_tokens > 0
        assert refill_rate > 0

        self.max_tokens = max_tokens
        self.refill_rate = refill_rate

        self.tokens = max_tokens
        self.last_refill_time = time.time()

        self.lock = Lock()

    def _refill(self) -> None:
        curr_time = time.time()
        diff_time = curr_time - self.last_refill_time

        missing_tokens = diff_time * self.refill_rate
        self.tokens = min(self.tokens + missing_tokens, self.max_tokens)

        self.last_refill_time = curr_time

    def allow_request(self) -> bool:
        with self.lock:
            self._refill()

            if self.tokens < 1.0:
                return False

            self.tokens -= 1.0
            return True

File: src/token_bucket.lua
local key = KEYS[1]

local max_tokens = ARGS[1]
local refill_rate = ARGS[2]
local curr_time = ARGS[3]
local requested_tokens = ARGS[4]

-- current bucket state

local data = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = data[1]
local last_refill = data[2]

if not tokens or not last_refill then
    tokens = max_tokens
    last_refill = now
else
    -- refill calculation
    missing_tokens = math.max(0, now - last_refill) * refill_rate
    tokens = math.min(max_tokens, tokens + missing_tokens)
    last_refill = now
end

-- deduct tokens for a request
if tokens >= requested_tokens then
    tokens = tokens - requested_tokens
    redis.call('HMSET', key, 'tokens', tokens, 'last_refill', last_refill)
    local.ttl = math.ceil(max_tokens/refill_rate * 3)
    redis.call('EXPIRE', key, ttl)
    return 1
else
    redis.call('HMSET', key, 'tokens', tokens, 'last_refill', last_refill)
    local.ttl = math.ceil(max_tokens/refill_rate * 3)
    redis.call('EXPIRE', key, ttl)
    return 0